persist-tool/persist-tool.c

```diff
@@ -0,0 +1,238 @@
+ * COPYING for details.
+ *
+ */
+
+#include "dump.h"
+#include "add.h"
+#include "generate.h"
+#include "persist-tool.h"
+#include "reloc.h"
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+
+#if HAVE_GETOPT_H
+#include <getopt.h>
+#endif
+
+static void
+load_state_handler_modules(GlobalConfig *cfg)
+{
+  plugin_context_init_instance(&cfg->plugin_context);
+  plugin_context_set_module_path(&cfg->plugin_context, (const gchar *)get_installation_path_for(SYSLOG_NG_MODULE_PATH));
+  plugin_load_candidate_modules(&cfg->plugin_context);
+}
+
+static gboolean
+persist_tool_start_state(PersistTool *self)
+{
+  if (!g_threads_got_initialized)
+    {
+      g_thread_init(NULL);
+    }
+  main_thread_handle = get_thread_id();
+  self->state = persist_state_new(self->persist_filename);
+  if (!persist_state_start(self->state))
+    {
+      fprintf(stderr, "Invalid persist file: %s\n", self->persist_filename);
+      persist_state_cancel(self->state);
+      persist_state_free(self->state);
+      self->state = NULL;
+      return FALSE;
+    }
+  return TRUE;
+}
+
+void
+persist_tool_revert_changes(PersistTool *self)
+{
+  if (self->state == NULL)
+    {
+      self->state = self->cfg->state;
+    }
+  persist_state_cancel(self->state);
+  persist_state_start(self->state);
+}
+
+PersistTool *
+persist_tool_new(gchar *persist_filename, PersistStateMode open_mode)
+{
+  PersistTool *self = g_new0(PersistTool, 1);
+  self->mode = open_mode;
+  self->persist_filename = g_strdup(persist_filename);
+  self->cfg = cfg_new(0);
+  if (!persist_tool_start_state(self))
+    {
+      persist_tool_free(self);
+      return NULL;
+    }
+  load_state_handler_modules(self->cfg);
+  return self;
+}
+
+void persist_tool_free(PersistTool *self)
+{
+  if (self->cfg)
+    {
+      cfg_free(self->cfg);
+    }
+  if (self->state)
+    {
+      if (self->mode != persist_mode_dump)
+        {
+          persist_state_commit(self->state);
+        }
+      else
+        {
+          persist_state_cancel(self->state);
+        }
+      persist_state_free(self->state);
+    }
+  g_free(self->persist_filename);
+  g_free(self);
+}
+
+static GOptionEntry dump_options[] =
+{
+  { NULL, 0, 0, G_OPTION_ARG_NONE, NULL, NULL }
+};
+
+static GOptionEntry add_options[] =
+{
+  { "output-dir", 'o', 0, G_OPTION_ARG_STRING, &persist_state_dir,"The directory where persist file is located.", "<directory>" },
+  { "persist-name", 'p', 0, G_OPTION_ARG_STRING, &persist_state_name, "The name of the persist file to generate. If not specified it will be the default syslog-ng.persist.", "<persist_name>"},
+  { NULL, 0, 0, G_OPTION_ARG_NONE, NULL, NULL }
+};
+
+static GOptionEntry generate_options[] =
+{
+  { "force", 'f', G_OPTION_FLAG_OPTIONAL_ARG, G_OPTION_ARG_NONE, &force_generate, "Overwrite the persist file if it already exists. WARNING: Use this option with care, persist-tool will not ask for confirmation", NULL},
```

Comment:
The `G_OPTION_FLAG_OPTIONAL_ARG` causes issue when executing this tool:
```
persist-tool generate -h

(process:27696): GLib-WARNING **: 07:00:53.213: ../../../../glib/goption.c:2393: ignoring no-arg, optional-arg or filename flags (32) on option of arg-type 0 in entry (null):force
Usage:
  persist-tool generate [OPTION?]
```

According to the documentation(https://developer.gnome.org/glib/stable/glib-Commandline-option-parser.html#GOptionFlags) with the above flag it still expects to receive a function, and not a boolean variable; you should simply have to use `0` instead of this flag.

Reply:
thanks, done